### DISCUSSION TITLE
Function to transform XML to JSON until Libraries are enabled

### DISCUSSION THREAD

**Discussion by cngi1** [Budibase/budibase#5445] [Help] [answered]
I have a 3rd party API that will only send XML data for parts of the implementation. Mostly it returns JSON but the use case I have needs the data that is only returned as XML.

Since we can't use any external libraries at the present time has anyone found two functions that can be included in a transformer.  1. to change the GET data  from XML to JSON for budibase to work with and 2. Convert JSON back to the XML for the PUT update?

Additionally what would be the best practice here to modify the data and ready it for the PUT back? Would local state.fieldA, B etc variables, local db temp storage or another Budibase way of working with the data be the best approach?

**Comment by cngi1** [comment]
any suggestions?

**Comment by melohagan** [comment]
Hi @cngi1 

There is no standard way in which the data you need will be returned from an xml source, however hopefully this answer can guide you in the right direction.

For this example, I used a Treasury Security API for pulling back a list of announced securities from an xhtml table. (https://www.treasurydirect.gov/webapis/webapisecurities.htm)
Endpoint: https://www.treasurydirect.gov/TA_WS/securities/announced?format=xhtml

**Transform Code**
```
function parseXmlToJson(xml) {
    const json = {};
    for (const res of xml.matchAll(/(?:<(\w*)(?:\s[^>]*)*>)((?:(?!<\1).)*)(?:<\/\1>)|<(\w*)(?:\s*)*\/>/gm)) {
        const key = res[1] || res[3];
	if (key === "tr" || key == "td") continue;
        const value = res[2] && parseXmlToJson(res[2]);
        json[key] = ((value && Object.keys(value).length) ? value : res[2]) || null;

    }
    return json;
}
function tableToJson(table) {
    let rows = [];
    let headers = [];
    for (const res of table.matchAll(/<th>(.*?)<\/th>/gm)) {
        headers.push(res[1])
    }
    let i = 0
    for (const res of table.matchAll(/<tr>(.*?)<\/tr>/gm)) {
	if (i === 0) {
	    i++;
            continue;
	}
	let j = 0
        let row = {}
	for (const tdres of res[1].matchAll(/<td>(.*?)<\/td>/gm)) {
	   row[headers[j]] = tdres[1]
           j++
	}
        rows.push(row)
    }
    return rows
}

let json = parseXmlToJson(data)
return tableToJson(json.html.body.table)
```

This code has been adapted from this StackOverflow post: https://stackoverflow.com/questions/1773550/convert-xml-to-json-and-back-using-javascript

**Result**

![Screenshot 2022-04-19 at 12 24 25](https://user-images.githubusercontent.com/101575380/163993099-889ec388-5a3a-49f3-8760-a335dee3eed0.png)

You can then click the X icons to remove fields you don't want, and also change the data types in the dropdowns. 

Depending on the format of the XML, you may only need the first *parseXmlToJson* function.

**Reply by cngi1** [reply]
The function above does not work for nested XML levels. FYI @melohagan 

using this sample xml.
`const xmlSample = '<tag>tag content</tag><tag2>another content</tag2><tag3><insideTag>inside content</insideTag><emptyTag/></tag3><tag4><insideTag><insidetag2><inside2a>IS2a</inside2a></insidetag2><insidetag3>IT3 stuff</insidetag3></insideTag><emptyTag/></tag4>';`

The function above does not convert <inside2a> </inside2a> to JSON. 

```
<tag>tag content</tag>
<tag2>another content</tag2>
<tag3>
	<insideTag>inside content</insideTag>
	<emptyTag/>
</tag3>
<tag4>
	<insideTag>
		<insidetag2>
			<inside2a>IS2a</inside2a>
		</insidetag2>
		<insidetag3>IT3 stuff</insidetag3>
	</insideTag>
	<emptyTag/>
</tag4>
```
```

`{
  tag: 'tag content',
  tag2: 'another content',
  tag3: { insideTag: 'inside content', emptyTag: null },
  tag4: {
    insideTag: { insidetag2: [Object], insidetag3: 'IT3 stuff' },
    emptyTag: null
  }
}`
```

**Reply by cngi1** [reply]
In testing the original function without table processing I was using 
console.log(parseXmlToJson(xmlSample));


`const xmlSample = '<tag>tag content</tag><tag2>tag2 content</tag2><tag3><insideTag0>insidetag0 content</insideTag0><emptyTag/></tag3><tag4><insideTag><insidetag2><inside2a>IS2a content</inside2a></insidetag2><insidetag3>IS3a content</insidetag3></insideTag><emptyTag/></tag4>';`


The output is
```
`{
  tag: 'tag content',
  tag2: 'another content',
  tag3: { insideTag: 'inside content', emptyTag: null },
  tag4: {
    insideTag: { insidetag2: [Object], insidetag3: 'IT3 stuff' },
    emptyTag: null
  }
}`

```
But after looking at the function I could not find a problem .
I finally tested with the following
```
let d = {};
d = parseXmlToJson(xmlSample, 0);
console.log(JSON.stringify(d));

Output
{"tag":"tag content","tag2":"tag2 content","tag3":{"insideTag0":"insidetag0 content","emptyTag":null},"tag4":{"insideTag":{"insidetag2":{"inside2a":"IS2a content"},"insidetag3":"IS3a content"},"emptyTag":null}}

```
The JSON object was actually correct from the beginning.


original
https://stackoverflow.com/questions/1773550/convert-xml-to-json-and-back-using-javascript

**Comment by cngi1** [comment]
Thank you very much for the example and the reference link. All the other example functions I was finding were vary large.
I will not need the table conversion but thank you again for the example. I will learn from it.

I do have to return PUT or POST as XML for this same end point after user updates form field(s).  Do you know of a json to xml function like this one?

**Reply by melohagan** [reply]
You're very welcome!

I found this json to xml that seemed well upvoted on StackOverflow: https://stackoverflow.com/questions/48788722/json-to-xml-using-javascript

With that said, you should be able to use bindings for variables within a fixed XML structure:
https://docs.budibase.com/docs/rest-bindings#body-example

So in this case you would click *raw (XML)* and enter in the XML with bindings, e.g.:

```
<root>
  <name>{{name}}</name>
  <age>{{age}}</age>
</root>
```

A little lower in that docs page, it demonstrates how to pass data from your app into your REST query:
https://docs.budibase.com/docs/rest-bindings#data-provider-applications-provider

**Reply by cngi1** [reply]
Thank you again. You have been a great help @melohagan

**Comment by cngi1** [comment, marked as answer]
Answers

https://github.com/Budibase/budibase/discussions/5445#discussioncomment-2592408

https://github.com/Budibase/budibase/discussions/5445#discussioncomment-2598913
